File: examples/maven-spring-compile/src/main/java/community/flock/wirespec/example/maven/custom/app/user/LiveUserClient.java
package community.flock.wirespec.example.maven.custom.app.user;

import community.flock.wirespec.example.maven.custom.app.common.WirespecSerializer;
import community.flock.wirespec.example.maven.custom.app.common.WirespecTransporter;
import community.flock.wirespec.generated.java.endpoint.DeleteUserByName;
import community.flock.wirespec.generated.java.endpoint.GetUserByName;
import community.flock.wirespec.generated.java.endpoint.GetUsers;
import community.flock.wirespec.generated.java.endpoint.PostUser;
import community.flock.wirespec.generated.java.endpoint.UploadImage;
import community.flock.wirespec.java.Wirespec;
import org.springframework.stereotype.Component;

import java.util.concurrent.CompletableFuture;

@Component
public class LiveUserClient implements UserClient {

    private final WirespecTransporter transporter;
    private final WirespecSerializer serializer;

    public LiveUserClient(final WirespecTransporter transporter,final  WirespecSerializer serializer) {
        this.transporter = transporter;
        this.serializer = serializer;
    }

    @Override
    public CompletableFuture<GetUsers.Response<?>> getUsers(final GetUsers.Request request) {
        return transporter.transport(GetUsers.Adapter.toRawRequest(this.serializer, request))
                .thenApplyAsync(res -> GetUsers.Adapter.fromRawResponse(this.serializer, res));
    }

    @Override
    public CompletableFuture<GetUserByName.Response<?>> getUserByName(final GetUserByName.Request request) {
        return transporter.transport(GetUserByName.Adapter.toRawRequest(this.serializer, request))
                .thenApplyAsync(res -> GetUserByName.Adapter.fromRawResponse(this.serializer, res));
    }

    @Override
    public CompletableFuture<PostUser.Response<?>> postUser(final PostUser.Request request) {
        return transporter.transport(PostUser.Adapter.toRawRequest(this.serializer, request))
                .thenApplyAsync(res -> PostUser.Adapter.fromRawResponse(this.serializer, res));
    }

    @Override
    public CompletableFuture<DeleteUserByName.Response<?>> deleteUserByName(final DeleteUserByName.Request request) {
        return transporter.transport(DeleteUserByName.Adapter.toRawRequest(this.serializer, request))
                .thenApplyAsync(res -> DeleteUserByName.Adapter.fromRawResponse(this.serializer, res));
    }

    @Override
    public CompletableFuture<UploadImage.Response<?>> uploadImage(UploadImage.Request request) {
        return transporter.transport(UploadImage.Adapter.toRawRequest(this.serializer, request))
                .thenApplyAsync(res -> UploadImage.Adapter.fromRawResponse(this.serializer, res));
    }
}

File: examples/gradle-ktor/src/main/kotlin/community/flock/wirespec/example/maven/custom/app/user/LiveUserClient.kt
package community.flock.wirespec.example.maven.custom.app.user

import community.flock.wirespec.example.maven.custom.app.common.Serialization
import community.flock.wirespec.example.maven.custom.app.common.WirespecClient
import community.flock.wirespec.generated.kotlin.endpoint.DeleteUserByName
import community.flock.wirespec.generated.kotlin.endpoint.GetUserByName
import community.flock.wirespec.generated.kotlin.endpoint.GetUsers
import community.flock.wirespec.generated.kotlin.endpoint.PostUser

interface UserClient :
    GetUsers.Handler,
    GetUserByName.Handler,
    PostUser.Handler,
    DeleteUserByName.Handler

class LiveUserClient(
    private val wirespec: WirespecClient,
) : UserClient {
    override suspend fun getUsers(request: GetUsers.Request) = with(GetUsers.Adapter) {
        request
            .let { toRawRequest(Serialization, it) }
            .let(wirespec::transport)
            .let { fromRawResponse(Serialization, it) }
    }

    override suspend fun getUserByName(request: GetUserByName.Request) = with(GetUserByName.Adapter) {
        request
            .let { toRawRequest(Serialization, it) }
            .let(wirespec::transport)
            .let { fromRawResponse(Serialization, it) }
    }

    override suspend fun postUser(request: PostUser.Request) = with(PostUser.Adapter) {
        request
            .let { toRawRequest(Serialization, it) }
            .let(wirespec::transport)
            .let { fromRawResponse(Serialization, it) }
    }

    override suspend fun deleteUserByName(request: DeleteUserByName.Request) = with(DeleteUserByName.Adapter) {
        request
            .let { toRawRequest(Serialization, it) }
            .let(wirespec::transport)
            .let { fromRawResponse(Serialization, it) }
    }
}
